File: CoderByte/algorithms/Tree_Traversal.go
package main

import (
	"fmt"
)

type Data rune

func (d Data) String() string {
	return fmt.Sprintf("%c", rune(d))
}

type Node struct {
	data        Data
	left, right *Node
}

func (root *Node) PreOrder(nodes []Data) []Data {
	if root == nil {
		return nodes
	}

	nodes = append(nodes, root.data)

	if root.left != nil {
		nodes = root.left.PreOrder(nodes)
	}

	if root.right != nil {
		nodes = root.right.PreOrder(nodes)
	}

	return nodes
}

func (root *Node) InOrder(nodes []Data) []Data {
	if root == nil {
		return nodes
	}

	if root.left != nil {
		nodes = root.left.InOrder(nodes)
	}

	nodes = append(nodes, root.data)

	if root.right != nil {
		nodes = root.right.InOrder(nodes)
	}

	return nodes
}

func (root *Node) PostOrder(nodes []Data) []Data {
	if root == nil {
		return nodes
	}

	if root.left != nil {
		nodes = root.left.PostOrder(nodes)
	}

	if root.right != nil {
		nodes = root.right.PostOrder(nodes)
	}

	nodes = append(nodes, root.data)

	return nodes
}

func (root *Node) LevelOrder(nodes []Data) []Data {
	queue := []*Node{root}

	for len(queue) > 0 {
		var n *Node

		// Pop
		n, queue = queue[0], queue[1:]

		nodes = append(nodes, n.data)

		if n.left != nil {
			queue = append(queue, n.left)
		}

		if n.right != nil {
			queue = append(queue, n.right)
		}
	}

	return nodes
}

func NewLeaf(data Data) *Node {
	return &Node{
		data: data,
	}
}

func NewNode(data Data, left, right *Node) *Node {
	return &Node{
		data:  data,
		left:  left,
		right: right,
	}
}

func main() {
	// create nodes
	n4 := NewLeaf('E')
	n3 := NewLeaf('D')
	n2 := NewLeaf('C')
	n1 := NewNode('B', n3, n4)
	root := NewNode('A', n1, n2)

	// empty list
	empty := []Data{}

	// do operations
	fmt.Println("Pre Order:", root.PreOrder(empty))
	fmt.Println("In Order:", root.InOrder(empty))
	fmt.Println("Post Order:", root.PostOrder(empty))
	fmt.Println("Level Order:", root.LevelOrder(empty))
}

File: CoderByte/algorithms/Tree_Traversal.py
class Node(object):
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right= None

#setup tree

root = Node("A")
n1   = Node("B")
n2   = Node("C")
n3   = Node("D")
n4   = Node("E")

#setup tree
root.left = n1
root.right= n2
n1.left = n3
n1.right= n4

def pre_order(root, nodes=[]):
    ''' root.data root.left root.right'''
    nodes.append(root.data)
    if root and root.left:pre_order(root.left, nodes)
    if root and root.right:pre_order(root.right, nodes)
    return nodes
seq = pre_order(root)
print("Pre order sequence is " ,end=' ')
print(*seq)

def in_order(root, nodes = []):
    ''' root.left root.data root.right'''
    if root and root.left: in_order(root.left, nodes)
    nodes.append(root.data)
    if root and root.right:in_order(root.right, nodes)
    return nodes

seq = in_order(root)
print("In order sequence is ", end=' ')
print(*seq)

def post_order(root, nodes=[]):
    ''' root.left root.right root.data'''
    if root and root.left:post_order(root.left, nodes)
    if root and root.right:post_order(root.right, nodes)
    nodes.append(root.data)
    return nodes

seq = post_order(root,[])
print("Post order sequence is ", end=" ")
print(*seq)

def level_order(root, nodes=[]):
    '''add root to queue,
       while queue pop queue
       add children of poped nodes to queue()'''
    queue = [root]
    while queue:
        node = queue.pop()
        nodes.append(node.data)
        if node.left != None: queue.insert(0, node.left)
        print
        if node.right!= None: queue.insert(0, node.right)
    return nodes

seq = level_order(root, [])
print("level order seq is :", end=' ')
print(*seq)